src/sources/mongodb_metrics/mod.rs

```diff
@@ -0,0 +1,1092 @@
+    CommandIsMasterMongoError { source: MongoError },
+    #[snafu(display("failed to parse `isMaster` response: {:?}", source))]
+    CommandIsMasterParseError { source: bson::de::Error },
+    #[snafu(display("failed to execute `buildInfo` command: {:?}", source))]
+    CommandBuildInfoMongoError { source: MongoError },
+    #[snafu(display("failed to parse `buildInfo` response: {:?}", source))]
+    CommandBuildInfoParseError { source: bson::de::Error },
+    #[snafu(display("only `Mongod` supported right now, current: `{:?}`", node_type))]
+    UnsupportedNodeType { node_type: NodeType },
+}
+
+#[derive(Debug)]
+enum CollectError {
+    Mongo(MongoError),
+    Bson(bson::de::Error),
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, Default)]
+#[serde(deny_unknown_fields)]
+struct MongoDBMetricsConfig {
+    endpoints: Vec<String>,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+    #[serde(default = "default_namespace")]
+    namespace: String,
+}
+
+#[derive(Debug)]
+struct MongoDBMetrics {
+    client: Client,
+    endpoint: String,
+    namespace: String,
+    tags: BTreeMap<String, String>,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+pub fn default_namespace() -> String {
+    "mongodb".to_string()
+}
+
+inventory::submit! {
+    SourceDescription::new::<MongoDBMetricsConfig>("mongodb_metrics")
+}
+
+impl_generate_config_from_default!(MongoDBMetricsConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "mongodb_metrics")]
+impl SourceConfig for MongoDBMetricsConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        mut shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        let mut interval = interval(Duration::from_secs(self.scrape_interval_secs)).map(|_| ());
+
+        let sources = try_join_all(
+            self.endpoints
+                .iter()
+                .map(|endpoint| MongoDBMetrics::new(endpoint, &self.namespace)),
+        )
+        .await?;
+
+        let mut out = out
+            .sink_map_err(|e| error!("error sending metric: {:?}", e))
+            .sink_compat();
+
+        Ok(Box::new(
+            async move {
+                loop {
+                    select! {
+                        _ = &mut shutdown => break,
+                        Some(()) = interval.next() => (),
+                        else => break,
+                    };
+
+                    let start = Instant::now();
+                    let metrics = join_all(sources.iter().map(|mongodb| mongodb.collect())).await;
+                    emit!(MongoDBMetricsCollectCompleted {
+                        start,
+                        end: Instant::now()
+                    });
+
+                    let mut stream = stream::iter(metrics).flatten().map(Event::Metric).map(Ok);
+                    if let Err(()) = out.send_all(&mut stream).await {
```

Comment:
Is this right? It seems like it should return a `SendError`.

Reply:
Yes, this is correct because we log errors (https://github.com/timberio/vector/pull/4500/files#diff-aaeacd9910e0a113de7912c22bcc5a074edb92fcc84b4b50481f3adf887bee55R143-R145):
```rust
        let mut out = out
            .sink_map_err(|e| error!("error sending metric: {:?}", e))
            .sink_compat();
```
:slightly_smiling_face:

Reply:
Aha :smile: I missed that it was reasigned above with `map_err`.